dgraphtest/cluster.go

```diff
@@ -0,0 +1,301 @@
+	}
+	return c.startContainer(c.alphas[id])
+}
+
+func (c *Cluster) startContainer(dc dnode) error {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	if err := c.dcli.ContainerStart(ctx, dc.cid(), types.ContainerStartOptions{}); err != nil {
+		return errors.Wrapf(err, "error starting container [%v]", dc.cname())
+	}
+	return nil
+}
+
+func (c *Cluster) Stop() error {
+	c.log("stopping cluster with prefix [%v]", c.conf.prefix)
+	for i := range c.alphas {
+		if err := c.StopAlpha(i); err != nil {
+			return err
+		}
+	}
+	for i := range c.zeros {
+		if err := c.StopZero(i); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Cluster) StopZero(id int) error {
+	if id >= c.conf.numZeros {
+		return fmt.Errorf("invalid id of zero: %v", id)
+	}
+	return c.stopContainer(c.zeros[id])
+}
+
+func (c *Cluster) StopAlpha(id int) error {
+	if id >= c.conf.numAlphas {
+		return fmt.Errorf("invalid id of alpha: %v", id)
+	}
+	return c.stopContainer(c.alphas[id])
+}
+
+func (c *Cluster) stopContainer(dc dnode) error {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	if err := c.dcli.ContainerStop(ctx, dc.cid(), &stopTimeout); err != nil {
+		return errors.Wrapf(err, "error stopping container [%v]", dc.cname())
+	}
+	return nil
+}
+
+func (c *Cluster) Cleanup() {
+	c.log("cleaning up cluster with prefix [%v]", c.conf.prefix)
+
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+
+	ro := types.ContainerRemoveOptions{RemoveVolumes: true, Force: true}
+	for _, aa := range c.alphas {
+		if err := c.dcli.ContainerRemove(ctx, aa.cid(), ro); err != nil {
+			c.log("error removing alpha [%v]: %v", aa.cname(), err)
+		}
+	}
+	for _, zo := range c.zeros {
+		if err := c.dcli.ContainerRemove(ctx, zo.cid(), ro); err != nil {
+			c.log("error removing zero [%v]: %v", zo.cname(), err)
+		}
+	}
+	if err := c.dcli.NetworkRemove(ctx, c.net.id); err != nil {
+		c.log("error removing network [%v]: %v", c.net.name, err)
+	}
+}
+
+func (c *Cluster) createContainer(dc dnode) (string, error) {
+	cmd := dc.cmd(c)
+	image := c.dgraphImage()
+	mts, err := dc.mounts(c.conf)
+	if err != nil {
+		return "", err
+	}
+
+	cconf := &container.Config{Cmd: cmd, Image: image, WorkingDir: dc.workingDir(), ExposedPorts: dc.ports()}
+	hconf := &container.HostConfig{Mounts: mts, PublishAllPorts: true}
+	networkConfig := &network.NetworkingConfig{
+		EndpointsConfig: map[string]*network.EndpointSettings{
+			c.net.name: {
+				Aliases:   []string{dc.cname(), dc.lname()},
+				NetworkID: c.net.id,
+			},
+		},
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+	resp, err := c.dcli.ContainerCreate(ctx, cconf, hconf, networkConfig, dc.cname())
+	if err != nil {
+		return "", errors.Wrapf(err, "error creating container %v", dc.cname())
+	}
+
+	return resp.ID, nil
+}
+
+func (c *Cluster) healthCheck() error {
```

Comment:
Is checking for health the same as checking if license is applied? Today we got an issue where we were checking for  health, but still license failed. Can we add a check for is license applied? Okay to be done in a separate diff.

Reply:
could you elaborate?